### PR TITLE
Add `Transaction` support to `FirestoreStore` and `FakeStore`

### PR DESCRIPTION
**Pull request by alexandrepiveteau**
This PR adds full support for transactions to `FirestoreStore`, and minimal support for transactions to `FakeStore` (transactions semantics are not respected, but instead document reads and writes are always performed and succeed in order).

### CODE REVIEW

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/store/Store.kt [hunk 1]
```diff
@@ -8,4 +8,16 @@ interface Store {
 
   /** Accesses a collection with the given [String] identifier. */
   fun collection(path: String): CollectionReference
+
+  /**
+   * Runs the give [Transaction] on the [Store], executing all the operations atomically. A
+   * [Transaction] must perform all its reads before performing all of its writes.
+   *
+   * This function may throw an [Exception] if a transaction can't be executed.
```
Comment (KurohanaJuri) [inline mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/store/Store.kt:16]: Can we add this information with a keyword, something like `@Exception` ? I don't if it's possible

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/store/Transaction.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
+package ch.epfl.sdp.mobile.infrastructure.persistence.store
+
+import kotlin.reflect.KClass
+
+/**
+ * A [Transaction] offers a way for clients to perform multiple [get] of documents, followed by some
+ * [set] and [update]. At most 500 documents may be modified per [Transaction].
```
[the remark this answers is not in the document]
Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/store/Transaction.kt:7]: Yes.

File: mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/store/Store.kt (hunk 1 above)
Comment (alexandrepiveteau) [reply mobile/src/main/java/ch/epfl/sdp/mobile/infrastructure/persistence/store/Store.kt:18]: Fixed in d1b4b2e0ff48edc3a1438212e25d1472214c6741.

**Review by matt989253 — APPROVED** [review]
Looks great to me! As discussed, we'll have to be careful with the `@UnsafeVariance`.

**Pull request merged**